base/array.jl

```diff
@@ -215,8 +215,8 @@ sizeof(a::Array) = Core.sizeof(a)
 
 function isassigned(a::Array, i::Int...)
```

Comment:
Yes, though this in particular seems hard to test :thinking: Is there a similar test for `getindex(::Array, ::Int)` we could steal?

Reply:
Started looking at this out of curiosity.

I found `test/boundscheck.jl` which executes `test/boundscheck_exec.jl` with different values for the `--check-bounds` flag. There are tests in there for `@inbounds` etc. for both `isassigned` and `getindex`.

I think something is maybe wrong with running the tests though. This test fails for me on master when I run it manually:
```
./julia --check-bounds=auto --startup-file=no --depwarn=error test/boundscheck_exec.jl
```

Same with the higher level script:
```
cd test/
../julia boundscheck.jl
```

It seems the `--check-bounds=auto` (default) case isn't executing during `make testall` or `make test[-revise]-boundscheck`. It isn't for me at least. I can throw `@test false` into those code paths and it won't hit it. As a result it seems a test failure may have snuck into the repo.

Reply:
M'kay that was happening because `Base.julia_cmd()` returns a command with the same flags as the running Julia process, and all the tests are run with `--check-bounds=yes` by default, so it was inheriting that. Need to add an explicit `--check-bounds=auto` in: https://github.com/JuliaLang/julia/blob/79e29e3cb9ef34f7ab613657c0f480d81b1402b7/test/boundscheck.jl#L5

Then the failing test runs which I guess is a whole other issue:
```
Test Failed at test/boundscheck_exec.jl:35
  Expression: A1_wrap() == 0
   Evaluated: 1 == 0
```

Reply:
Yes, that failure seems unrelated (though nonetheless bad), thanks for looking into it! That indeed seems a bit broken. `A1_wrap` wraps `A1_inbounds`, which I'd expect to be `0`, since it has its own internal `@inbounds` that eliminates the boundscheck. Also interesting that `A1_noinline()` does not seem to be tested.

Please open an issue for this, or a PR for the `--check-bounds=auto` change so there's an explicit place for discussion.

Reply:
Made https://github.com/JuliaLang/julia/issues/48099

Reply:
The tests are fixed on master now, so I was looking again at changing `@inline` -> `@propagate_inbounds`. If I understand correctly how the propagation works (big "if"), I don't think it makes any concrete difference at the moment. None of the functions directly called by `isassigned` has any `@boundscheck`. It looks like I'd have to propagate through quite a few more layers before making it to something like `getindex` where maybe it would matter. Thoughts?

Reply:
`@boundscheck` in a caller is not required for `@propagate_inbounds` to work, only `@inbounds` is. As an example:

```julia
function foo(arr, e)
      s = 0
      @inbounds for i in 1:e
          s += bar(arr, i)
      end
      s
end

function bar(arr, idx)
       @boundscheck 1 <= idx <= length(arr)
       z = arr[idx]
       z^2 + 20z
end
```

would be something where the `@boundscheck` can be removed, since the caller guarantees via the presence of the `@inbounds` that they checked that all array indices will be inbounds. I think that's where it could make a difference with `@inline` vs. `@propagate_inbounds`, since there's not enough information in `foo` itself for the compiler to figure it out by itself. Indeed, using `@inline` is not enough there:

```julia
julia> function bar(arr, idx)
              @inline
              @boundscheck 1 <= idx <= length(arr)
              z = arr[idx]
              z^2 + 20z
       end
bar (generic function with 1 method)


julia> @code_native foo(Int[], 1)
# [...]
.LBB0_7:                                # %oob
	mov	rsp, rsi
	movabs	rax, offset ijl_bounds_error_ints
	mov	edx, 1
	mov	qword ptr [rsi], r8
	call	rax
.Lfunc_end0:
	.size	julia_foo_707, .Lfunc_end0-julia_foo_707
	.cfi_endproc
                                        # -- End function
	.section	".note.GNU-stack","",@progbits
```

Reply:
Thanks, I misunderstood the problem `@propagate_inbounds` was solving. Playing around with your foo/bar example cleared it up.

I don't seem to get the same behavior with `isassigned`. The compiler seems to be happy in this hacky example with `@inline`, the call to `checkbounds` is removed:

```julia
function isassigned2(a::Array, i::Int...)
    @inline
    @boundscheck checkbounds(Bool, a, i...) || return false
    ii = (Base._sub2ind(size(a), i...) % UInt) - 1
    ccall(:jl_array_isassigned, Cint, (Any, UInt), a, ii) == 1
end

function isassigned_wrap(a, i)
    r = 0
    @inbounds for ii in 1:i
        r += (isassigned2(a, i) ? ii : 0)
    end
    return r > 0
end

@test isassigned_wrap(Int[], 2)
```

Even still I can submit the one liner change as a PR. Maybe the compiler is fine with this case but would be unhappy without `@propagate_inbounds` in some other case. If I could find one of those cases I could maybe write a test to go along.

Reply:
I have found a case where `@propagate_inbounds` matters, but sadly only in auto-vectorizable code https://github.com/eschnett/SIMD.jl/pull/102

I can share the code if you'd like to take a look, but it's a pretty dense function (similar to the one I have in #48236, but written with SIMD.jl)

Reply:
Sure I'll take a look